File: hardware/push/bridge.js
// Synthteam hardware bridge for the Ableton Push 3 — all JS.
//
// Connects the physical Push directly to the Synthteam server as a
// dedicated `push-join` hardware client and runs the Push station's
// activities on the hardware:
//
//   push-path    trace the lit vector across the pads (server-checked;
//                pad events are forwarded as hardware-events)
//   push-defend  DEFEND THE MOTHERSHIP — the server declares the
//                activity with missile parameters, the bridge runs the
//                minigame locally: missiles climb the pad columns,
//                pressing their pad intercepts them, impacts wound the
//                saucer on screen. Survive until the task deadline and
//                the server completes it; if the local hull reaches
//                zero the bridge reports push-defend-failed, which the
//                server treats like an expired order.
//
// The screen shows mission state throughout: lobby/countdown/score/
// integrity/deadlines, the vector path or the mothership, and a HULL
// BREACH flash whenever the ship takes damage. The button row above
// the pads (CC 102-109) is the defend activity's hull bar.
//
// Run:  bun bridge.js  [--server ws://127.0.0.1:4179/ws]

import { drawText, fillRect, textWidth } from "./font.js";
import { HEIGHT, PAD_COLORS, Push3, WIDTH } from "./push3.js";

// Push palette index + screen RGB for each PushPathColor the game uses.
const PATH_COLORS = {
  violet: [PAD_COLORS.violet, [170, 80, 255]],
  cyan: [PAD_COLORS.cyan, [0, 220, 220]],
  amber: [PAD_COLORS.amber, [255, 160, 0]],
  lime: [PAD_COLORS.lime, [150, 255, 40]],
};

const HULL_FLASH_MS = 1500;
const EXPLOSION_MS = 220;
const IMPACT_FLASH_MS = 700;

// Our saucer, filling the full width of the screen: gray upper dome
// (cropped by the top edge), a rim of glowing windows, and a yellow
// bottom hemisphere with landing lights. 48 columns: 1 hull, 4 dome,
// 2 rim windows, 3 under-dome lights, 5 yellow hemisphere.
const SHIP_SPRITE = [
  "0".repeat(16) + "4".repeat(16) + "0".repeat(16),
  "0".repeat(12) + "4".repeat(24) + "0".repeat(12),
  "0".repeat(4) + "1".repeat(40) + "0".repeat(4),
  "11" + "2211".repeat(11) + "11",
  "1".repeat(48),
  "0".repeat(6) + "1".repeat(36) + "0".repeat(6),
  "0".repeat(14) + "5".repeat(20) + "0".repeat(14),
  "0".repeat(17) + "55335533553355" + "0".repeat(17),
  "0".repeat(20) + "5".repeat(8) + "0".repeat(20),
];
const SHIP_COLORS = {
  1: [105, 115, 135],
  2: [255, 220, 140],
  3: [255, 240, 180],
  4: [70, 78, 95],
  5: [235, 190, 40],
};
const SHIP_SCALE_X = WIDTH / SHIP_SPRITE[0].length;
const SHIP_SCALE_Y = 10;

const serverArg = process.argv.indexOf("--server");
const serverUrl =
  serverArg >= 0 ? process.argv[serverArg + 1] : "ws://127.0.0.1:4179/ws";

const push = await Push3.open();
console.log("push bridge: hardware linked");

let connected = false;
let phase = { kind: "lobby" };
let task = null;
let activeGridSize = 0;
let missionFlashUntil = 0; // full-screen breach flash (path mode only)
let lastIntegrity = null;
let socket = null;
const litPathPads = new Set();

// Local state for the running push-defend activity, or null.
let defend = null;

function initDefend(defendTask) {
  clearPadGrid();
  defend = {
    taskId: defendTask.id,
    missiles: [],   // {col, y: float 0..8 climbing}
    explosions: [], // {col, row, until, killed}
    wounds: [],     // ship damage {x, y} in sprite pixels
    hull: defendTask.hull,
    kills: 0,
    impactUntil: 0,
    failed: false,
    lastSpawn: performance.now(),
    lastTick: performance.now(),
  };
}

function endDefend() {
  if (defend === null) {
    return;
  }
  defend = null;
  clearPadGrid();
}

function clearPadGrid() {
  for (let x = 0; x < 8; x++) {
    for (let y = 0; y < 8; y++) {
      push.setPad(x, y, PAD_COLORS.off);
    }
  }
  litPathPads.clear();
  litDefendPads.clear();
}

connect();

function connect() {
  socket = new WebSocket(serverUrl);
  socket.onopen = () => {
    connected = true;
    socket.send(JSON.stringify({ type: "push-join" }));
    console.log("push bridge: joined", serverUrl);
  };
  socket.onmessage = (event) => {
    const message = JSON.parse(String(event.data));
    if (message.type === "push-state") {
      applyState(message.state);
    }
  };
  socket.onclose = () => {
    if (connected) {
      console.log("push bridge: server lost, retrying");
    }
    connected = false;
    phase = { kind: "lobby" };
    task = null;
    activeGridSize = 0;
    lastIntegrity = null;
    endDefend();
    lightTask(null);
    setTimeout(connect, 1000);
  };
  socket.onerror = () => {};
}

setInterval(() => {
  if (connected) {
    socket.send(JSON.stringify({ type: "ping" }));
  }
}, 30_000);

push.on("pad", ({ x, y, down, velocity }) => {
  if (!connected) {
    return;
  }
  // During a defend activity the pads are the interceptor battery,
  // handled locally; otherwise they belong to the server.
  if (defend !== null && task?.kind === "push-defend") {
    if (down && !defend.failed) {
      interceptAt(x, y);
    }
    return;
  }
  if (x >= activeGridSize || y >= activeGridSize) {
    return;
  }
  socket.send(
    JSON.stringify({
      type: "hardware-event",
      event: {
        kind: "push-pad",
        point: { x, y },
        phase: down ? "down" : "up",
        velocity,
      },
    }),
  );
});

function applyState(state) {
  phase = state.phase;
  task = state.task;
  activeGridSize = state.activeGridSize;

  if (task?.kind === "push-defend" && phase.kind === "playing") {
    if (defend === null || defend.taskId !== task.id) {
      initDefend(task);
    }
  } else {
    endDefend();
  }

  const integrity = phase.kind === "playing" ? phase.integrity : null;
  if (
    defend === null &&
    integrity !== null &&
    lastIntegrity !== null &&
    integrity < lastIntegrity
  ) {
    missionFlashUntil = performance.now() + HULL_FLASH_MS;
  }
  lastIntegrity = integrity;

  if (defend === null) {
    lightTask(task);
  }
}

/** Show the playable region, then overlay the current path or corner targets. */
function lightTask(activeTask) {
  const wanted = new Map();
  for (let x = 0; x < activeGridSize; x++) {
    for (let y = 0; y < activeGridSize; y++) {
      wanted.set(y * 8 + x, PAD_COLORS.blue);
    }
  }
  if (activeTask?.kind === "push-path") {
    const [led] = PATH_COLORS[activeTask.color] ?? [PAD_COLORS.white];
    activeTask.path.forEach((point, index) => {
      if (index < activeTask.progress) {
        return; // already traced
      }
      const key = point.y * 8 + point.x;
      wanted.set(key, index === activeTask.progress ? PAD_COLORS.white : led);
    });
  } else if (activeTask?.kind === "push-corners") {
    const farEdge = activeTask.gridSize - 1;
    for (const point of [
      { x: 0, y: 0 },
      { x: farEdge, y: 0 },
      { x: 0, y: farEdge },
      { x: farEdge, y: farEdge },
    ]) {
      const pressed = activeTask.pressed.some(
        (candidate) => candidate.x === point.x && candidate.y === point.y,
      );
      wanted.set(
        point.y * 8 + point.x,
        pressed ? PAD_COLORS.green : PAD_COLORS.white,
      );
    }
  }
  for (const key of litPathPads) {
    if (!wanted.has(key)) {
      push.setPad(key % 8, Math.floor(key / 8), PAD_COLORS.off);
    }
  }
  litPathPads.clear();
  for (const [key, color] of wanted) {
    push.setPad(key % 8, Math.floor(key / 8), color);
    litPathPads.add(key);
  }
}

// ---- Defend activity simulation ----

function interceptAt(x, y) {
  const now = performance.now();
  const hit = defend.missiles.findIndex(
    (missile) => missile.col === x && Math.floor(missile.y) === y,
  );
  if (hit >= 0) {
    defend.missiles.splice(hit, 1);
    defend.kills++;
    defend.explosions.push({
      col: x, row: y, until: now + EXPLOSION_MS, killed: true,
    });
  } else {
    defend.explosions.push({ col: x, row: y, until: now + 100, killed: false });
  }
}

function woundShip() {
  for (let attempt = 0; attempt < 30; attempt++) {
    const y = Math.floor(Math.random() * SHIP_SPRITE.length);
    const x = Math.floor(Math.random() * SHIP_SPRITE[0].length);
    if (SHIP_SPRITE[y][x] !== "0" &&
        !defend.wounds.some((hole) => hole.x === x && hole.y === y)) {
      defend.wounds.push({ x, y });
      if (Math.random() < 0.6 && attempt < 29) {
        continue; // impacts usually tear more than one hole
      }
      return;
    }
  }
}

function tickDefend() {
  if (defend === null || task?.kind !== "push-defend" || defend.failed) {
    return;
  }
  const now = performance.now();
  const dt = (now - defend.lastTick) / 1000;
  defend.lastTick = now;

  // No fresh launches once the survival clock has run out
  const clockRunning = Date.now() < task.deadlineAt;
  if (clockRunning && now - defend.lastSpawn > task.spawnIntervalMs) {
    const busyCols = new Set(
      defend.missiles
        .filter((missile) => missile.y < 1.5)
        .map((missile) => missile.col),
    );
    const freeCols = [...Array(8).keys()].filter((col) => !busyCols.has(col));
    if (freeCols.length > 0) {
      const col = freeCols[Math.floor(Math.random() * freeCols.length)];
      defend.missiles.push({ col, y: 0.01 });
      defend.lastSpawn = now;
    }
  }

  for (const missile of defend.missiles) {
    missile.y += task.missileSpeed * dt;
  }

  const impacts = defend.missiles.filter((missile) => missile.y >= 8);
  if (impacts.length > 0) {
    defend.missiles = defend.missiles.filter((missile) => missile.y < 8);
    defend.hull -= impacts.length;
    defend.impactUntil = now + IMPACT_FLASH_MS;
    for (const _ of impacts) {
      woundShip();
    }
    if (defend.hull <= 0) {
      defend.hull = 0;
      defend.failed = true;
      defend.missiles = [];
      if (connected) {
        socket.send(
          JSON.stringify({
            type: "hardware-event",
            event: { kind: "push-defend-failed" },
          }),
        );
      }
    }
  }

  defend.explosions = defend.explosions.filter(
    (explosion) => explosion.until > now,
  );
}

// ---- Pad LEDs for the defend battle ----

const litDefendPads = new Map();

function renderDefendPads() {
  if (defend === null) {
    return;
  }
  const wanted = new Map();
  if (!defend.failed) {
    for (const missile of defend.missiles) {
      wanted.set(Math.floor(missile.y) * 8 + missile.col, PAD_COLORS.red);
    }
    for (const explosion of defend.explosions) {
      wanted.set(
        explosion.row * 8 + explosion.col,
        explosion.killed ? PAD_COLORS.white : PAD_COLORS.amber,
      );
    }
  }
  for (const [key, color] of litDefendPads) {
    if (wanted.get(key) !== color && !wanted.has(key)) {
      push.setPad(key % 8, Math.floor(key / 8), PAD_COLORS.off);
      litDefendPads.delete(key);
    }
  }
  for (const [key, color] of wanted) {
    if (litDefendPads.get(key) !== color) {
      push.setPad(key % 8, Math.floor(key / 8), color);
      litDefendPads.set(key, color);
    }
  }
}

// ---- Button backlights: CC 102-109 is the defend hull bar ----

const BUTTON_ROW = [102, 103, 104, 105, 106, 107, 108, 109];
const litButtons = new Map();

function renderButtons() {
  const now = performance.now();
  for (const [index, cc] of BUTTON_ROW.entries()) {
    let color = PAD_COLORS.off;
    if (defend !== null) {
      if (defend.failed || now < defend.impactUntil) {
        color = PAD_COLORS.red;
      } else if (index < defend.hull) {
        color = PAD_COLORS.yellow;
      }
    }
    if (litButtons.get(cc) !== color) {
      push.setButton(cc, color);
      litButtons.set(cc, color);
    }
  }
}

// ---- Screen ----

const buffer = Buffer.alloc(WIDTH * HEIGHT * 4);
buffer.fill(255);

function centered(text, y, scale, color) {
  drawText(buffer, WIDTH, text,
    (WIDTH - textWidth(text, scale)) / 2, y, scale, color);
}

function drawShip(now) {
  const shipY = -SHIP_SCALE_Y * 1.5 + Math.sin(now / 450) * 4;
  const flashing = defend !== null && now < defend.impactUntil;
  const windowPulse = 0.7 + 0.3 * Math.sin(now / 300);
  const wounds = defend?.wounds ?? [];
  const critical = defend !== null && defend.hull <= 2;

  for (let row = 0; row < SHIP_SPRITE.length; row++) {
    for (let col = 0; col < SHIP_SPRITE[row].length; col++) {
      const cell = SHIP_SPRITE[row][col];
      if (cell === "0") {
        continue;
      }
      let color = SHIP_COLORS[cell];
      if (cell === "2") {
        color = color.map((c) => Math.round(c * windowPulse));
      }
      if (cell === "3" && Math.sin(now / 120 + col) > 0) {
        color = [140, 100, 20]; // landing lights blink
      }
      if (wounds.some((hole) => hole.x === col && hole.y === row)) {
        color = Math.sin(now / 90 + col * 3 + row) > 0.3
          ? [255, 120, 20]  // embers glow in the wounds
          : [30, 20, 18];
      } else if (flashing) {
        color = [255, 255, 255];
      } else if (critical) {
        color = color.map((c) => Math.round(c * 0.75));
      }
      fillRect(buffer, WIDTH,
        col * SHIP_SCALE_X, shipY + row * SHIP_SCALE_Y,
        SHIP_SCALE_X, SHIP_SCALE_Y, ...color);
    }
  }
  return shipY + SHIP_SPRITE.length * SHIP_SCALE_Y;
}

function renderDefendScreen(now) {
  const impact = now < defend.impactUntil;
  fillRect(buffer, WIDTH, 0, 0, WIDTH, HEIGHT,
    impact ? 90 + Math.round(50 * Math.sin(now / 25)) : 8, 12, impact ? 12 : 24);

  const shipBottom = drawShip(now);

  if (defend.failed) {
    centered("MOTHERSHIP HIT - ACTIVITY FAILED", 104, 4, [224, 85, 85]);
    return;
  }

  for (const missile of defend.missiles) {
    const x = ((missile.col + 0.5) / 8) * WIDTH;
    const headY = HEIGHT - (missile.y / 8) * (HEIGHT - shipBottom);
    for (let segment = 0; segment < 5; segment++) {
      const fade = 1 - segment / 5;
      fillRect(buffer, WIDTH,
        x - 4, headY + segment * 9, 8, 7,
        Math.round(255 * fade), Math.round(40 * fade), Math.round(30 * fade));
    }
  }
  for (const explosion of defend.explosions.filter((e) => e.killed)) {
    const x = ((explosion.col + 0.5) / 8) * WIDTH;
    const grow = 1 - (explosion.until - now) / EXPLOSION_MS;
    const radius = 8 + grow * 22;
    const y = HEIGHT - ((explosion.row + 0.5) / 8) * (HEIGHT - shipBottom);
    fillRect(buffer, WIDTH, x - radius, y - radius / 2,
      radius * 2, radius, 255, Math.round(220 - 120 * grow), 60);
  }

  drawText(buffer, WIDTH, "HULL", 16, HEIGHT - 44, 2, [140, 150, 173]);
  fillRect(buffer, WIDTH, 16, HEIGHT - 24, 220, 14, 60, 66, 82);
  fillRect(buffer, WIDTH, 18, HEIGHT - 22, 216 * (defend.hull / task.hull), 10,
    235, 190, 40);
  const secondsLeft = Math.max(0, (task.deadlineAt - Date.now()) / 1000);
  centered("DEFEND THE MOTHERSHIP", HEIGHT - 56, 2, [140, 150, 173]);
  centered(`${Math.ceil(secondsLeft)}S`, HEIGHT - 36, 4,
    secondsLeft < 4 ? [255, 230, 0] : [200, 205, 215]);
  drawText(buffer, WIDTH, `SCORE ${phase.score}`,
    WIDTH - textWidth(`SCORE ${phase.score}`, 4) - 16, HEIGHT - 32, 4,
    [127, 212, 255]);
  if (impact) {
    centered("HULL BREACH!", HEIGHT - 76, 5, [255, 255, 255]);
  }
}

function renderPathScreen(now) {
  fillRect(buffer, WIDTH, 0, 0, WIDTH, HEIGHT, 8, 12, 24);

  const cell = 17;
  const gridX = 12;
  const gridY = 8;
  for (let gx = 0; gx < 8; gx++) {
    for (let gy = 0; gy < 8; gy++) {
      const active = gx < activeGridSize && gy < activeGridSize;
      fillRect(buffer, WIDTH,
        gridX + gx * cell, gridY + (7 - gy) * cell,
        cell - 3, cell - 3, ...(active ? [20, 26, 42] : [5, 7, 10]));
    }
  }
  if (task?.kind === "push-path") {
    const [, rgb] = PATH_COLORS[task.color] ?? [0, [255, 255, 255]];
    task.path.forEach((point, index) => {
      const traced = index < task.progress;
      const isNext = index === task.progress;
      const color = traced ? [60, 66, 82] : isNext ? [255, 255, 255] : rgb;
      fillRect(buffer, WIDTH,
        gridX + point.x * cell, gridY + (7 - point.y) * cell,
        cell - 3, cell - 3, ...color);
    });
    drawText(buffer, WIDTH, `${task.color} VECTOR`, 170, 12, 4, rgb);
    drawText(buffer, WIDTH,
      `${task.progress}/${task.path.length} TRACED`, 170, 52, 3,
      [200, 205, 215]);
  } else if (task?.kind === "push-corners") {
    const farEdge = task.gridSize - 1;
    for (const point of [
      { x: 0, y: 0 },
      { x: farEdge, y: 0 },
      { x: 0, y: farEdge },
      { x: farEdge, y: farEdge },
    ]) {
      const pressed = task.pressed.some(
        (candidate) => candidate.x === point.x && candidate.y === point.y,
      );
      fillRect(buffer, WIDTH,
        gridX + point.x * cell, gridY + (7 - point.y) * cell,
        cell - 3, cell - 3, ...(pressed ? [70, 214, 140] : [255, 255, 255]));
    }
    drawText(buffer, WIDTH, "CORNERS!", 170, 12, 5, [150, 255, 40]);
    drawText(buffer, WIDTH, `${task.pressed.length}/4 HIT`, 170, 54, 3,
      [200, 205, 215]);
  } else {
    drawText(buffer, WIDTH, "STANDBY", 170, 26, 6, [140, 150, 173]);
  }
  drawText(buffer, WIDTH, `SCORE ${phase.score}`, 430, 12, 4, [127, 212, 255]);
  if (phase.combo > 1) {
    drawText(buffer, WIDTH, `COMBO X${phase.combo}`, 430, 52, 3, [255, 230, 0]);
  }
  const integrity = Math.max(0, Math.min(100, phase.integrity));
  const barColor = integrity > 50 ? [70, 214, 140]
    : integrity > 25 ? [255, 160, 0] : [224, 85, 85];
  drawText(buffer, WIDTH, "INTEGRITY", 660, 12, 2, [140, 150, 173]);
  fillRect(buffer, WIDTH, 660, 34, 280, 18, 60, 66, 82);
  fillRect(buffer, WIDTH, 662, 36, 276 * (integrity / 100), 14, ...barColor);
  drawText(buffer, WIDTH,
    `LEVEL ${phase.level}  ${phase.levelObjectivesCompleted}/${phase.levelObjectiveTarget}`,
    660, 66, 2,
    [140, 150, 173]);
  if (task !== null) {
    const orderLeft = Math.max(0, task.deadlineAt / 1000 - Date.now() / 1000);
    drawText(buffer, WIDTH, `INPUT EXPIRES ${orderLeft.toFixed(1)}S`, 170, 100, 3,
      orderLeft < 3 ? [224, 85, 85] : [140, 150, 173]);
  }
}

function renderScreen() {
  const now = performance.now();

  if (defend === null && now < missionFlashUntil) {
    const pulse = Math.round(160 + 95 * Math.sin(now / 33));
    fillRect(buffer, WIDTH, 0, 0, WIDTH, HEIGHT, pulse, 0, 0);
    const text = "HULL BREACH!";
    centered(text, (HEIGHT - 7 * 12) / 2 + Math.cos(now / 16) * 5, 12,
      [255, 255, 255]);
    void push.display.rgba(buffer);
    return;
  }

  switch (phase.kind) {
    case "lobby": {
      fillRect(buffer, WIDTH, 0, 0, WIDTH, HEIGHT, 8, 12, 24);
      drawText(buffer, WIDTH, "SYNTHTEAM", 20, 24, 11, [127, 212, 255]);
      drawText(buffer, WIDTH,
        `PUSH STATION - ${connected ? "LINKED TO CENTRAL" : "WAITING FOR CENTRAL"}`,
        22, 118, 3, connected ? [140, 150, 173] : [224, 85, 85]);
      break;
    }
    case "countdown": {
      fillRect(buffer, WIDTH, 0, 0, WIDTH, HEIGHT, 8, 12, 24);
      const remaining = Math.max(0, Math.ceil(phase.endsAt / 1000 - Date.now() / 1000));
      centered(String(remaining), (HEIGHT - 7 * 18) / 2, 18, [255, 230, 0]);
      break;
    }
    case "playing": {
      if (defend !== null && task?.kind === "push-defend") {
        renderDefendScreen(now);
      } else {
        renderPathScreen(now);
      }
      break;
    }
    case "game-over": {
      fillRect(buffer, WIDTH, 0, 0, WIDTH, HEIGHT, 8, 12, 24);
      const survived = phase.reason === "survived";
      const text = survived ? "MISSION SURVIVED" : "SHIP LOST";
      const color = survived ? [70, 214, 140] : [224, 85, 85];
      centered(text, 26, 9, color);
      centered(`FINAL SCORE ${phase.score}`, 110, 4, [200, 205, 215]);
      break;
    }
  }
  void push.display.rgba(buffer);
}

const timer = setInterval(() => {
  tickDefend();
  renderDefendPads();
  renderButtons();
  renderScreen();
}, 33);

process.on("SIGINT", () => {
  clearInterval(timer);
  for (const cc of BUTTON_ROW) {
    push.setButton(cc, PAD_COLORS.off);
  }
  void push.close().then(() => process.exit(0));
});

File: hardware/push/bridge.py
"""Synthteam hardware bridge for the Ableton Push 3.

Connects the physical Push directly to the Synthteam server as a
dedicated `push-join` hardware client — an alternative to driving the
Push through Chrome's Web MIDI in the central console. Because it talks
to the Push over USB it can also use the 960x160 display, which Web
MIDI cannot: mission phase, the live vector path, score/integrity, and
a red HULL BREACH flash whenever the ship takes damage.

No grid learning is needed: the Push 3 pad grid is fixed at notes
36-99, bottom-left origin, matching the server's (x, y) convention of
the learned-corner setup (bottom-left = (0, 0)).

Run:  uv run bridge.py  [--server ws://127.0.0.1:4179/ws]
"""

import argparse
import asyncio
import json
import threading
import time

import aiohttp
import mido
import numpy as np
from PIL import Image, ImageDraw, ImageFont

from push3_display import HEIGHT, WIDTH, Push3Display

PORT_NAME = "Ableton Push 3 Live Port"
PAD_LOW, PAD_HIGH = 36, 99
LED_OFF, LED_WHITE, LED_RED = 0, 3, 6

# Push palette index + screen RGB for each PushPathColor the game uses.
PATH_COLORS = {
    "violet": (79, (170, 80, 255)),
    "cyan": (34, (0, 220, 220)),
    "amber": (10, (255, 160, 0)),
    "lime": (18, (150, 255, 40)),
}

HULL_FLASH_SECONDS = 1.5


def pad_note(x, y):
    return PAD_LOW + y * 8 + x


class State:
    """Shared between the MIDI callback thread, render thread and asyncio."""

    def __init__(self):
        self.lock = threading.Lock()
        self.connected = False
        self.phase = {"kind": "lobby"}
        self.task = None            # PushPathTask dict from the server
        self.flash_until = 0.0      # HULL BREACH deadline
        self.last_integrity = None


class PushBridge:
    def __init__(self):
        self.state = State()
        self.display = Push3Display()
        self.midi_out = mido.open_output(PORT_NAME)
        self.event_queue = asyncio.Queue()
        self.loop = asyncio.get_running_loop()
        self.lit_notes = set()
        self.midi_in = mido.open_input(PORT_NAME, callback=self.on_midi)

    # ---- Pads (MIDI callback thread) -> server events ----

    def on_midi(self, msg):
        if msg.type not in ("note_on", "note_off"):
            return
        if not PAD_LOW <= msg.note <= PAD_HIGH:
            return
        x = (msg.note - PAD_LOW) % 8
        y = (msg.note - PAD_LOW) // 8
        down = msg.type == "note_on" and msg.velocity > 0
        event = {
            "type": "hardware-event",
            "event": {
                "kind": "push-pad",
                "point": {"x": x, "y": y},
                "phase": "down" if down else "up",
                "velocity": msg.velocity,
            },
        }
        self.loop.call_soon_threadsafe(self.event_queue.put_nowait, event)

    # ---- Server state -> pad LEDs ----

    def apply_state(self, push_state):
        with self.state.lock:
            self.state.phase = push_state.get("phase", {"kind": "lobby"})
            self.state.task = push_state.get("task")
            integrity = (
                self.state.phase.get("integrity")
                if self.state.phase.get("kind") == "playing"
                else None
            )
            if (
                integrity is not None
                and self.state.last_integrity is not None
                and integrity < self.state.last_integrity
            ):
                self.state.flash_until = time.monotonic() + HULL_FLASH_SECONDS
            self.state.last_integrity = integrity
            task = self.state.task
        self.light_path(task)

    def light_path(self, task):
        """Light the remaining path in its color; next expected pad is white."""
        wanted = {}
        if task is not None:
            led, _ = PATH_COLORS.get(task["color"], (LED_WHITE, None))
            progress = task["progress"]
            for index, point in enumerate(task["path"]):
                if index < progress:
                    continue  # already traced
                note = pad_note(point["x"], point["y"])
                wanted[note] = LED_WHITE if index == progress else led
        for note in self.lit_notes - set(wanted):
            self.midi_out.send(mido.Message("note_on", note=note, velocity=LED_OFF))
        for note, velocity in wanted.items():
            self.midi_out.send(mido.Message("note_on", note=note, velocity=velocity))
        self.lit_notes = set(wanted)

    # ---- Screen (render thread) ----

    def render_loop(self):
        font = ImageFont.load_default(size=26)
        small = ImageFont.load_default(size=18)
        big = ImageFont.load_default(size=80)
        huge = ImageFont.load_default(size=96)
        while True:
            with self.state.lock:
                phase = self.state.phase
                task = self.state.task
                connected = self.state.connected
                flash_left = self.state.flash_until - time.monotonic()
            try:
                self.display.send_image(
                    self.render(phase, task, connected, flash_left,
                                font, small, big, huge))
            except Exception:
                pass
            time.sleep(0.03)

    def render(self, phase, task, connected, flash_left, font, small, big, huge):
        now = time.monotonic()

        if flash_left > 0:
            pulse = 0.55 + 0.45 * np.sin(now * 30)
            image = Image.new("RGB", (WIDTH, HEIGHT), (int(160 + 95 * pulse), 0, 0))
            draw = ImageDraw.Draw(image)
            text = "HULL BREACH!!!"
            tw = draw.textlength(text, font=huge)
            tx = (WIDTH - tw) / 2 + np.sin(now * 47) * 8
            ty = (HEIGHT - 96) / 2 + np.cos(now * 61) * 6
            draw.text((tx + 4, ty + 4), text, font=huge, fill=(60, 0, 0))
            draw.text((tx, ty), text, font=huge, fill=(255, 255, 255))
            return image

        image = Image.new("RGB", (WIDTH, HEIGHT), (8, 12, 24))
        draw = ImageDraw.Draw(image)
        kind = phase.get("kind", "lobby")

        if kind == "lobby":
            draw.text((20, 30), "SYNTHTEAM", font=big, fill=(127, 212, 255))
            status = "linked to central" if connected else "waiting for central..."
            draw.text((22, 120), f"PUSH STATION  ·  {status}",
                      font=small, fill=(140, 150, 173))
        elif kind == "countdown":
            remaining = max(0.0, phase["endsAt"] / 1000 - time.time())
            text = f"{remaining:.0f}"
            tw = draw.textlength(text, font=huge)
            draw.text(((WIDTH - tw) / 2, 25), text, font=huge, fill=(255, 230, 0))
        elif kind == "playing" and task is not None:
            _, rgb = PATH_COLORS.get(task["color"], (LED_WHITE, (255, 255, 255)))
            # Mini grid mirroring the pads, path + progress
            cell, ox, oy = 17, 12, 8
            for gx in range(8):
                for gy in range(8):
                    x0 = ox + gx * cell
                    y0 = oy + (7 - gy) * cell
                    draw.rectangle([x0, y0, x0 + cell - 3, y0 + cell - 3],
                                   fill=(20, 26, 42))
            for index, point in enumerate(task["path"]):
                x0 = ox + point["x"] * cell
                y0 = oy + (7 - point["y"]) * cell
                traced = index < task["progress"]
                is_next = index == task["progress"]
                color = ((60, 66, 82) if traced
                         else (255, 255, 255) if is_next else rgb)
                draw.rectangle([x0, y0, x0 + cell - 3, y0 + cell - 3], fill=color)
            # Status block
            draw.text((170, 8), f"{task['color'].upper()} VECTOR",
                      font=font, fill=rgb)
            draw.text((170, 44), f"{task['progress']}/{len(task['path'])} traced",
                      font=small, fill=(200, 205, 215))
            score = phase.get("score", 0)
            combo = phase.get("combo", 0)
            draw.text((420, 8), f"SCORE {score}", font=font, fill=(127, 212, 255))
            if combo > 1:
                draw.text((420, 44), f"combo x{combo}", font=small,
                          fill=(255, 230, 0))
            # Integrity bar
            integrity = max(0, min(100, phase.get("integrity", 100)))
            draw.text((660, 8), "INTEGRITY", font=small, fill=(140, 150, 173))
            bar_color = ((70, 214, 140) if integrity > 50
                         else (255, 160, 0) if integrity > 25 else (224, 85, 85))
            draw.rectangle([660, 34, 940, 52], outline=(60, 66, 82), width=2)
            draw.rectangle([662, 36, 662 + int(276 * integrity / 100), 50],
                           fill=bar_color)
            # Mission clock + task deadline
            remaining = max(0.0, phase["endsAt"] / 1000 - time.time())
            draw.text((660, 66), f"mission {remaining:.0f}s", font=small,
                      fill=(140, 150, 173))
            order_left = max(0.0, task["deadlineAt"] / 1000 - time.time())
            draw.text((170, 100), f"order expires {order_left:.1f}s",
                      font=small,
                      fill=(224, 85, 85) if order_left < 3 else (140, 150, 173))
        elif kind == "game-over":
            survived = phase.get("reason") == "survived"
            text = "MISSION SURVIVED" if survived else "SHIP LOST"
            color = (70, 214, 140) if survived else (224, 85, 85)
            tw = draw.textlength(text, font=big)
            draw.text(((WIDTH - tw) / 2, 25), text, font=big, fill=color)
            score = phase.get("score", 0)
            sub = f"final score {score}"
            sw = draw.textlength(sub, font=font)
            draw.text(((WIDTH - sw) / 2, 115), sub, font=font, fill=(200, 205, 215))

        return image

    # ---- WebSocket ----

    async def sender(self, ws):
        while True:
            await ws.send_json(await self.event_queue.get())

    async def run(self, server_url):
        threading.Thread(target=self.render_loop, daemon=True).start()
        async with aiohttp.ClientSession() as session:
            while True:
                try:
                    async with session.ws_connect(server_url, heartbeat=30) as ws:
                        await ws.send_json({"type": "push-join"})
                        with self.state.lock:
                            self.state.connected = True
                        print("linked to", server_url, flush=True)
                        send_task = asyncio.ensure_future(self.sender(ws))
                        try:
                            async for raw in ws:
                                if raw.type != aiohttp.WSMsgType.TEXT:
                                    continue
                                msg = json.loads(raw.data)
                                if msg.get("type") == "push-state":
                                    self.apply_state(msg["state"])
                        finally:
                            send_task.cancel()
                except aiohttp.ClientError:
                    pass
                with self.state.lock:
                    self.state.connected = False
                    self.state.phase = {"kind": "lobby"}
                    self.state.task = None
                    self.state.last_integrity = None
                self.light_path(None)
                await asyncio.sleep(1.0)


async def amain():
    parser = argparse.ArgumentParser()
    parser.add_argument("--server", default="ws://127.0.0.1:4179/ws")
    args = parser.parse_args()
    bridge = PushBridge()
    try:
        await bridge.run(args.server)
    finally:
        bridge.light_path(None)


def main():
    try:
        asyncio.run(amain())
    except KeyboardInterrupt:
        pass


if __name__ == "__main__":
    main()
